lib/bap_types/bap_ir.ml

```diff
@@ -143,18 +143,60 @@ type sub = {
 
 
 type path = int array
-  [@@deriving bin_io, compare, sexp]
-
-
-type program = {
-  subs  : sub term array;
-  paths : path Tid.Table.t;
-} [@@deriving bin_io, fields, sexp]
-
-let compare_program x y =
-  let compare x y = [%compare:sub term array] x y in
-  compare x.subs y.subs
+[@@deriving bin_io, compare, sexp]
+
+module Program : sig
+  type t = private {
+    subs  : sub term array;
+    paths : path Tid.Table.t;
+  } [@@deriving bin_io, compare, fields, sexp]
+
+  val create : ?subs:sub term array -> ?paths:path Tid.Table.t -> unit -> t
+end = struct
+  type t = {
+    subs  : sub term array;
+    paths : path Tid.Table.t;
+  } [@@deriving bin_io, fields, sexp]
+
+  let mangle_name addr tid name =
+    match addr with
+    | Some a ->
+      sprintf "%s@%s" name @@
+      Bap_bitvector.string_of_value ~hex:true a
+    | None -> sprintf "%s%%%s" name (Tid.to_string tid)
+
+  let fix_names subs =
+    let names = String.Table.create () in
+    Array.iter subs ~f:(fun s ->
+        let name = s.self.name in
+        let tid = s.tid in
+        Hashtbl.change names name ~f:(function
+            | None -> Some tid
+            | Some tid' ->
+              if Tid.(tid' < tid) then Some tid
```

Comment:
So this policy prioritizes a subroutine with a smaller tid, though it doesn't give the priority to the last chosen name.